[SIL] Represent copies to and from @sil_weak values.

nate-chandler opened the pull request:
Introduce two new instructions:
- `strong_copy_weak_value`
- `weak_copy_value`

These are transformed by AddressLowering to, respectively:
- `load_weak`
- `store_weak`

The effect of `strong_copy_weak_value %o : $@sil_weak Optional<T>` is to
increment the strong reference count of the value in the `.some` if any.

The effect of `weak_copy_value %o : $Optional<T>` is to increment the weak
reference count of the value in the `.some` case if any.

nate-chandler committed:
[AddressLowering] Handle weak copies.

Lower the `strong_copy_weak_value` and `weak_copy_value` to `load_weak`
and `store_weak` respectively.

nate-chandler committed:
[SIL] Added strong_copy_weak_value.

The new instruction unwraps an `@sil_weak` box and produces an owned
value. It is only legal in opaque values mode and is transformed by
`AddressLowering` to `load_weak`.

nate-chandler committed:
[SIL] Added weak_copy_value.

The new instruction wraps a value in a `@sil_weak` box and produces an
owned value. It is only legal in opaque values mode and is transformed
by `AddressLowering` to `store_weak`.

nate-chandler committed:
[OpaqueValues] Emit weak copies.

In opaque values mode, emit the new weak copy instructions to convert as
follows:

strong_copy_weak_value: `@owned $sil_weak T?` -> `@owned $T?`
weak_copy_value: `@owned $T?` -> `@owned $@sil_weak T?`

Doing so is necessary in opaque values mode where it is needed to deal
with weak values directly rather than indirectly via `load_weak`s and
`store_weak`s.

nate-chandler committed:
[SIL] Doc: Tweaked load_weak comment.

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test windows platform

nate-chandler commented:
@swift-ci please test linux platform

nate-chandler commented:
@swift-ci please test linux platform

nate-chandler commented:
@swift-ci please test linux platform

nate-chandler commented:
@swift-ci please test linux platform

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test macos platform

nate-chandler commented:
@swift-ci please test

atrick reviewed (approved):
Makes sense!
In a follow-up PR, you my want to assert in SIL verification that the 'ref_to' instructions cannot be used for address-only values.

nate-chandler committed:
[SIL] Verifier: Disable opaque values for ref_to.

The ref_to_* and *_to_ref instructions must not produce or take as their
operands values of address-only type.  The AddressLowering pass would
trap on encountering such illegal instructions already.  Enforce the
invariant in the verifier.

nate-chandler commented:
@atrick Added that verification in https://github.com/apple/swift/pull/67690/commits/2396f97410eade706a865adf98e940f85b80dbad since there was a conflict resolve anyway.

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test
